File: app/src/main/java/com/zxcx/zhizhe/ui/MainActivity.java
package com.zxcx.zhizhe.ui;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.RadioButton;
import android.widget.RadioGroup;

import com.zxcx.zhizhe.R;
import com.zxcx.zhizhe.event.ChangeNightModeEvent;
import com.zxcx.zhizhe.event.ClassifyClickRefreshEvent;
import com.zxcx.zhizhe.event.HomeClickRefreshEvent;
import com.zxcx.zhizhe.mvpBase.BaseActivity;
import com.zxcx.zhizhe.ui.card.card.cardDetails.CardDetailsActivity;
import com.zxcx.zhizhe.ui.card.cardBag.CardBagActivity;
import com.zxcx.zhizhe.ui.classify.ClassifyFragment;
import com.zxcx.zhizhe.ui.home.HomeFragment;
import com.zxcx.zhizhe.ui.loginAndRegister.login.LoginActivity;
import com.zxcx.zhizhe.ui.my.MyFragment;
import com.zxcx.zhizhe.ui.welcome.WebViewActivity;
import com.zxcx.zhizhe.utils.Constants;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class MainActivity extends BaseActivity {

    @BindView(R.id.home_fragment_content)
    FrameLayout mHomeFragmentContent;
    @BindView(R.id.home_tab_home)
    RadioButton mHomeTabHome;
    @BindView(R.id.home_tab_classify)
    RadioButton mHomeTabAll;
    @BindView(R.id.home_tab_my)
    RadioButton mHomeTabMy;
    @BindView(R.id.rg_main)
    RadioGroup mRgMain;

    private Fragment mCurrentFragment = new Fragment();
    private HomeFragment mHomeFragment = new HomeFragment();
    private ClassifyFragment mClassifyFragment = new ClassifyFragment();
    private MyFragment mMyFragment = new MyFragment();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        EventBus.getDefault().register(this);

        mHomeTabHome.performClick();

        Intent intent = getIntent();
        //判断是否点击了广告或通知
        gotoLoginActivity(intent);
        gotoADActivity(intent);
        gotoNotificationActivity(intent);
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        //避免恢复视图状态
    }

    @Override
    public void recreate() {
        Intent intent = new Intent();
        setIntent(intent);
        super.recreate();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        EventBus.getDefault().unregister(this);
    }

    @OnClick({R.id.home_tab_home, R.id.home_tab_classify, R.id.home_tab_my})
    public void onViewClicked(View view) {
        switch (view.getId()) {
            case R.id.home_tab_home:
                switchFragment(mHomeFragment);
                break;
            case R.id.home_tab_classify:
                switchFragment(mClassifyFragment);
                break;
            case R.id.home_tab_my:
                switchFragment(mMyFragment);
                break;
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMessageEvent(ChangeNightModeEvent event) {
        this.recreate();
    }

    private void switchFragment(Fragment newFragment) {

        FragmentManager fm = getSupportFragmentManager();
        FragmentTransaction transaction = fm.beginTransaction();

        if (mCurrentFragment == newFragment) {
            if (newFragment == mHomeFragment) {
                EventBus.getDefault().post(new HomeClickRefreshEvent());
            } else if (newFragment == mClassifyFragment) {
                EventBus.getDefault().post(new ClassifyClickRefreshEvent());
            }
        } else {
            if (newFragment.isAdded()) {
                //.setCustomAnimations(R.anim.fragment_anim_left_in,R.anim.fragment_anim_right_out)
                transaction.hide(mCurrentFragment).show(newFragment).commitAllowingStateLoss();
            } else {
                transaction.hide(mCurrentFragment).add(R.id.home_fragment_content, newFragment).commitAllowingStateLoss();
            }
            mCurrentFragment = newFragment;
        }
    }

    private void gotoLoginActivity(Intent intent) {
        if (intent.getBooleanExtra("isFirst", false)) {
            Intent intent1 = new Intent(this, LoginActivity.class);
            startActivity(intent1);
        }
    }

    private void gotoADActivity(Intent intent) {
        if (intent.getBooleanExtra("hasAd", false)) {
            Intent intent1 = new Intent(this, WebViewActivity.class);
            intent1.putExtra("title", intent.getStringExtra("title"));
            intent1.putExtra("url", intent.getStringExtra("url"));
            startActivity(intent1);
        }
    }

    private void gotoNotificationActivity(Intent intent) {
        Bundle bundle = intent.getBundleExtra("push");
        if (bundle != null) {
            String type = bundle.getString("type");
            Intent detailIntent = new Intent();

            switch (type) {
                case Constants.PUSH_TYPE_CARD_BAG:
                    detailIntent.setClass(this, CardBagActivity.class);
                    break;
                case Constants.PUSH_TYPE_CARD:
                    detailIntent.setClass(this, CardDetailsActivity.class);
                    break;
                case Constants.PUSH_TYPE_AD:
                    detailIntent.setClass(this, WebViewActivity.class);
                    break;
            }
            detailIntent.putExtras(bundle);
            startActivity(detailIntent);
        }
    }
}

File: app/src/main/java/com/zxcx/zhizhe/ui/MainActivity.kt
package com.zxcx.zhizhe.ui

import android.app.SharedElementCallback
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.View
import com.zxcx.zhizhe.R
import com.zxcx.zhizhe.event.ChangeNightModeEvent
import com.zxcx.zhizhe.event.HomeClickRefreshEvent
import com.zxcx.zhizhe.mvpBase.BaseActivity
import com.zxcx.zhizhe.ui.article.HomeArticleFragment
import com.zxcx.zhizhe.ui.card.HomeCardFragment
import com.zxcx.zhizhe.ui.loginAndRegister.login.LoginActivity
import com.zxcx.zhizhe.ui.my.MyFragment
import com.zxcx.zhizhe.ui.my.creation.CreationAgreementDialog
import com.zxcx.zhizhe.ui.my.creation.newCreation.CreationEditorActivity
import com.zxcx.zhizhe.ui.my.creation.newCreation.CreationEditorLongActivity
import com.zxcx.zhizhe.ui.my.pastelink.PasteLinkActivity
import com.zxcx.zhizhe.ui.my.writer_status_writer
import com.zxcx.zhizhe.ui.rank.RankFragment
import com.zxcx.zhizhe.ui.welcome.WebViewActivity
import com.zxcx.zhizhe.utils.SVTSConstants
import com.zxcx.zhizhe.utils.SharedPreferencesUtil
import com.zxcx.zhizhe.utils.startActivity
import com.zxcx.zhizhe.widget.PublishDialog
import kotlinx.android.synthetic.main.activity_main.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

/**
 * 首页
 */

class MainActivity : BaseActivity() {

    private var mCurrentFragment = Fragment()
    private var mHomeCardFragment = HomeCardFragment()
    private var mHomeArticleFragment = HomeArticleFragment()
    private var mRankFragment = RankFragment()
    private var mMyFragment: MyFragment? = MyFragment()
    private var mIsReenter = false

    //////
//    var advList:ArrayList<AdInfo> = ArrayList()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        EventBus.getDefault().register(this)
        initShareElement()
        val intent = intent
        //判断是否点击了广告或通知
        gotoLoginActivity(intent)
        gotoADActivity(intent)
        if (intent.getBooleanExtra("isNight", false)) {
            home_tab_my.performClick()
        } else {
            home_tab_card.performClick()
        }

//        showFirstDialog()
//        val adManager = AdManager(this,advList)
//        adManager.setOverScreen(true)
//                .setPageTransformer(DepthPageTransformer())
//                .setOnImageClickListener { view, advInfo ->
//                    toastShow("get AD")
//                }
//
//        adManager.showAdDialog(AdConstant.ANIM_DOWN_TO_UP)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        //避免恢复视图状态
    }

    override fun recreate() {
        val intent = Intent()
        intent.putExtra("isNight", true)
        setIntent(intent)
        try {//避免重启太快 恢复
            val fragmentTransaction = supportFragmentManager.beginTransaction()
            for (fragment in supportFragmentManager.fragments) {
                fragmentTransaction.remove(fragment)
            }
            fragmentTransaction.commitAllowingStateLoss()
        } catch (e: Exception) {
        }

        super.recreate()
    }

    override fun onDestroy() {
        EventBus.getDefault().unregister(this)
        super.onDestroy()
    }

    override fun setListener() {
        super.setListener()

//        iv_home_creation.setOnClickListener {
//            if (checkLogin()) {
//                when (SharedPreferencesUtil.getInt(SVTSConstants.writerStatus, 0)) {
//                    writer_status_writer -> {
//                        //选择模板
//                        mActivity.startActivity(TemplateCardActivity::class.java) {}
//                    }
//                    else -> {
//                        val dialog = CreationAgreementDialog()
//                        dialog.mListener = {
//                            mActivity.startActivity(TemplateCardActivity::class.java){}
//                        }
//                        dialog.show(mActivity.supportFragmentManager,"")
//                    }
//                }
//            }
//        }

        //暂时注释，可能版本仍需确认
//        iv_home_creation.setOnClickListener {
//            if (checkLogin()) {
//                when (SharedPreferencesUtil.getInt(SVTSConstants.writerStatus, 0)) {
//                    writer_status_writer -> {
//                        //创作界面
//                        mActivity.startActivity(CreationEditorActivity::class.java) {}
//                    }
//                    else -> {
//                        val dialog = CreationAgreementDialog()
//                        dialog.mListener = {
//                            mActivity.startActivity(CreationEditorActivity::class.java) {}
//                        }
//                        dialog.show(mActivity.supportFragmentManager, "")
//                    }
//                }
//            }
//        }

        iv_home_creation.setOnClickListener {
            val mHomeDialog = PublishDialog(this)
            mHomeDialog.setFabuClickListener {
                //                toastShow("开始创作")
                if (checkLogin()) {
                    when (SharedPreferencesUtil.getInt(SVTSConstants.writerStatus, 0)) {
                        writer_status_writer -> {
                            //创作界面
                            mActivity.startActivity(CreationEditorActivity::class.java) {}
                        }

                        else -> {
                            val dialog = CreationAgreementDialog()
                            dialog.mListener = {
                                mActivity.startActivity(CreationEditorActivity::class.java) {}
                            }
                            dialog.show(mActivity.supportFragmentManager, "")
                        }
                    }
                }
                mHomeDialog.outDia()
            }

            mHomeDialog.setShenDuClickListener {
                if (checkLogin()) {
                    when (SharedPreferencesUtil.getInt(SVTSConstants.writerStatus, 0)) {
                        writer_status_writer -> {
                            //创作界面
                            mActivity.startActivity(CreationEditorLongActivity::class.java) {}
                        }

                        else -> {
                            val dialog = CreationAgreementDialog()
                            dialog.mListener = {
                                mActivity.startActivity(CreationEditorLongActivity::class.java) {}
                            }
                            dialog.show(mActivity.supportFragmentManager, "")
                        }
                    }
                }
                mHomeDialog.outDia()
            }

            mHomeDialog.setHuishouClickListener {

                if (checkLogin()) {
                    when (SharedPreferencesUtil.getInt(SVTSConstants.writerStatus, 0)) {
                        writer_status_writer -> {
                            //一键转载
                            mActivity.startActivity(PasteLinkActivity::class.java) {}
                        }

                        else -> {
                            val dialog = CreationAgreementDialog()
                            dialog.mListener = {
                                mActivity.startActivity(PasteLinkActivity::class.java) {}
                            }
                            dialog.show(mActivity.supportFragmentManager,"")
                        }
                    }
                }

//                mActivity.startActivity(PasteLinkActivity::class.java) {}

                mHomeDialog.outDia()
            }

            mHomeDialog.show()
        }

        home_tab_card.setOnClickListener { switchFragment(mHomeCardFragment) }
        home_tab_article.setOnClickListener { switchFragment(mHomeArticleFragment) }
        home_tab_rank.setOnClickListener {
            switchFragment(mRankFragment)
        }
        home_tab_my.setOnClickListener { switchFragment(mMyFragment) }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(event: ChangeNightModeEvent) {
        this.recreate()
    }

    private fun switchFragment(newFragment: Fragment?) {

        val fm = supportFragmentManager
        val transaction = fm.beginTransaction()

        if (mCurrentFragment === newFragment) {
            if (newFragment === mHomeCardFragment) {
                EventBus.getDefault().post(HomeClickRefreshEvent())
            }
        } else {
            if (newFragment!!.isAdded) {
                //.setCustomAnimations(R.anim.fragment_anim_left_in,R.anim.fragment_anim_right_out)
                transaction.hide(mCurrentFragment).show(newFragment).commitAllowingStateLoss()
            } else {
                transaction.hide(mCurrentFragment).add(R.id.home_fragment_content, newFragment).commitAllowingStateLoss()
            }
            mCurrentFragment = newFragment
        }
    }

    private fun gotoLoginActivity(intent: Intent) {
        if (intent.getBooleanExtra("needLogin", false)) {
            val intent1 = Intent(this, LoginActivity::class.java)
            startActivity(intent1)
        }
    }

    private fun gotoADActivity(intent: Intent) {
        if (intent.getBooleanExtra("hasAd", false)) {
            val intent1 = Intent(this, WebViewActivity::class.java)
            intent1.putExtra("title", intent.getStringExtra("title"))
            intent1.putExtra("url", intent.getStringExtra("url"))
            startActivity(intent1)
        }
    }

    override fun onActivityReenter(resultCode: Int, data: Intent?) {
        super.onActivityReenter(resultCode, data)
        if (mCurrentFragment == mHomeCardFragment) {
            mIsReenter = true
            postponeEnterTransition()
            mHomeCardFragment.onActivityReenter()
        }
    }

    fun getSharedView(names: MutableList<String>): MutableMap<String, View>? {
        return if (mCurrentFragment == mHomeCardFragment) {
            mHomeCardFragment.getSharedView(names)
        } else {
            null
        }
    }

    private fun initShareElement() {
        setExitSharedElementCallback(mExitCallback)
    }

    private val mExitCallback = object : SharedElementCallback() {

        override fun onMapSharedElements(names: MutableList<String>, sharedElements: MutableMap<String, View>) {
            if (mIsReenter) {
                sharedElements.clear()
                sharedElements.putAll(getSharedView(names) ?: mutableMapOf())
                mIsReenter = false
            }
        }
    }

    ////////////////////////////
//    private fun showFirstDialog() {
//        val adInfo = AdInfo()
//        adInfo.activityImg = "https://raw.githubusercontent.com/yipianfengye/android-adDialog/master/images/testImage1.png"
//        advList.add(adInfo)
//    }
}
